rotkehlchen/chain/evm/decoding/decoder.py

```diff
@@ -187,10 +187,16 @@ def _recursively_initialize_decoders(
                         )
                         continue
 
-                    results.address_mappings.update(self.decoders[class_name].addresses_to_decoders())  # noqa: E501
+                    decoders = self.decoders[class_name].addresses_to_decoders()
+                    decoders_with_counterparty = {}
+                    first_counterparty = self.decoders[class_name].counterparties()[0]
+                    for address, decoders_by_address in decoders.items():
+                        decoders_with_counterparty[address] = (decoders_by_address, first_counterparty)  # noqa: E501
+
+                    results.address_mappings.update(decoders_with_counterparty)
                     results.event_rules.extend(self.decoders[class_name].decoding_rules())
-                    results.token_enricher_rules.extend(self.decoders[class_name].enricher_rules())
-                    results.post_decoding_rules.extend(self.decoders[class_name].post_decoding_rules())  # noqa: E501
+                    results.token_enricher_rules.extend([(self.decoders[class_name].enricher_rules(), first_counterparty)])  # noqa: E501
+                    results.post_decoding_rules += self.decoders[class_name].post_decoding_rules()  # noqa: E501
```

Comment:
Okay so here I don't understand what you are doing. You need to add some comments.

1. addresses_to_decoders is not the decoders. A decoder is the class. Name the variable accordingly. If you meana single decoding function, that is a decoding rule.
2. Why `first_counterparty`? What is the significance? A decoder can have multiple counterparties. I don't understand what you are doing here.
3. Wow I really don't get what you are doing here in those lines. Are you trying to map counterparties to post-decoding rules? Why not just dothis as part of each module's `post_decoding_rules()` function? IT's the responsiility of the module to give you the map. Not trying to find it out post-facto by some obscure logic.